tasks/2048/src/main.rs

```diff
@@ -1,379 +1,567 @@
-                        RKey::Up => Some(Key::Up),
-                        RKey::Down => Some(Key::Down),
-                        RKey::Left => Some(Key::Left),
-                        RKey::Right => Some(Key::Right),
-                        _ => None,
-                    }
+    fn is_move_available(&self) -> bool {
+        for x in 0..4 {
+            for y in 0..4 {
+                if self[(x, y)].is_none() {
+                    return true;
                 }
-                Err(e) => panic!("{}", e),
-                _ => None,
+                if x < 3 && self[(x + 1, y)] == self[(x, y)] {
+                    return true;
+                };
+                if x > 0 && self[(x - 1, y)] == self[(x, y)] {
+                    return true;
+                };
+                if y < 3 && self[(x, y + 1)] == self[(x, y)] {
+                    return true;
+                };
+                if y > 0 && self[(x, y - 1)] == self[(x, y)] {
+                    return true;
+                };
             }
         }
+        false
+    }
 
-        fn draw_grid(&self, grid: [[usize; 4]; 4]) {
-            let x = 0;
-            let y = 2;
-            let width = 36;
-            let height = 18;
-            let cell_width: usize = width / grid.len();
-            let cell_height: usize = height / grid[0].len();
-            for i in 0..grid.len() {
-                let x_coord = x + i * cell_width + i;
-                for j in 0..grid[0].len() {
-                    let y_coord = y + j * cell_height + j;
-                    let x_text_offset = cell_width / 2;
-                    let y_text_offset = cell_height / 2;
-                    let num: String = format!("{}", grid[i][j]);
-                    let x_text_offset = x_text_offset - num.len() / 4;
-
-                    let rect = Rectangle {
-                        x: x_coord,
-                        y: y_coord,
-                        w: cell_width,
-                        h: cell_height,
-                    };
-                    self.draw_rectangle(&rect, Color::Black, Color::White, Color::Black);
-                    if num != "0" {
-                        self.rustbox.print(x_coord + x_text_offset,
-                                           y_coord + y_text_offset,
-                                           rustbox::RB_BOLD,
-                                           Color::White,
-                                           Color::Black,
-                                           &num);
-                    }
-                }
-            }
+    fn spawn_tile(&mut self) -> Result<(), GridFullError> {
+        if !self.is_move_available() {
+            return Err(GridFullError);
         }
 
-        fn present(&self) {
-            self.rustbox.present();
+        loop {
+            let (x, y) = rand::random::<(usize, usize)>();
+            let rand_tile = &mut self[(x % GRID_DIMENSION, y % GRID_DIMENSION)];
```

Comment:
This should either use `rand::Rng::gen_range` or mention how `% 4` avoids modulo bias.